File: zena_mode/profiler.py
import time
import functools
import logging
import uuid
from collections import deque

logger = logging.getLogger("ZenAIProfiler")


class PerformanceMonitor:
    """
    Singleton class to track and hold performance metrics.
    """

    _instance = None

    def __new__(cls):
        """New."""
        if cls._instance is not None:
            return

        cls._instance = super(PerformanceMonitor, cls).__new__(cls)
        cls._instance.metrics = {
            "llm_tps": deque(maxlen=50),  # Tokens per second
            "llm_ttft": deque(maxlen=50),  # Time to first token
            "rag_retrieval": deque(maxlen=50),  # ms for RAG fetch
            "expert_latency": deque(maxlen=50),  # ms for swarm consensus
            "traces": {},  # Trace ID -> detailed logs
        }
        return cls._instance

    def start_trace(self) -> str:
        """Generates a new trace ID for a request batch."""
        trace_id = str(uuid.uuid4())[:8]
        self.metrics["traces"][trace_id] = []
        return trace_id

    def log_trace(self, trace_id: str, message: str):
        """Append a message to a specific trace."""
        if trace_id not in self.metrics["traces"]:
            return

        self.metrics["traces"][trace_id].append(f"[{time.strftime('%H:%M:%S')}] {message}")
        logger.debug(f"[Trace:{trace_id}] {message}")

    def add_metric(self, key: str, value: float):
        if key not in self.metrics:
            return

        self.metrics[key].append(value)
        logger.debug(f"[Metric] {key}: {value:.2f}")

    def get_averages(self):
        return {k: (sum(v) / len(v) if v else 0) for k, v in self.metrics.items()}


monitor = PerformanceMonitor()


def profile_execution(name: str):
    """
    Decorator to measure execution time of a function.
    Automatically logs results to PerformanceMonitor.
    """

    def decorator(func):
        """Decorator."""

        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            """Wrapper."""
            start = time.time()
            result = func(*args, **kwargs)
            duration_ms = (time.time() - start) * 1000

            # Map function names to internal metric keys
            if "rag" in name.lower():
                monitor.add_metric("rag_retrieval", duration_ms)
            elif "expert" in name.lower():
                monitor.add_metric("expert_latency", duration_ms)

            logger.info(f"[Profiler] {name} took {duration_ms:.1f}ms")
            return result

        return wrapper

    return decorator


def profile_async_execution(name: str):
    """
    Async version of the execution profiler.
    """

    def decorator(func):
        """Decorator."""

        @functools.wraps(func)
        async def wrapper(*args, **kwargs):
            """Wrapper."""
            start = time.time()
            result = await func(*args, **kwargs)
            duration_ms = (time.time() - start) * 1000

            if "rag" in name.lower():
                monitor.add_metric("rag_retrieval", duration_ms)

            logger.info(f"[Profiler] {name} took {duration_ms:.1f}ms")
            return result

        return wrapper

    return decorator

File: rust_output/src/profiler.rs
use anyhow::{Result, Context};
use std::collections::HashMap;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

pub static MONITOR: std::sync::LazyLock<PerformanceMonitor> = std::sync::LazyLock::new(|| Default::default());

/// Singleton class to track and hold performance metrics.
#[derive(Debug, Clone)]
pub struct PerformanceMonitor {
}

impl PerformanceMonitor {
    /// New.
    pub fn __new__(&self, cls: String) -> () {
        // New.
        if cls._instance.is_some() {
            return;
        }
        cls._instance = r#super(PerformanceMonitor, cls).__new__(cls);
        cls._instance.metrics = HashMap::from([("llm_tps".to_string(), deque(/* maxlen= */ 50)), ("llm_ttft".to_string(), deque(/* maxlen= */ 50)), ("rag_retrieval".to_string(), deque(/* maxlen= */ 50)), ("expert_latency".to_string(), deque(/* maxlen= */ 50)), ("traces".to_string(), HashMap::new())]);
        cls._instance
    }
    /// Generates a new trace ID for a request batch.
    pub fn start_trace(&mut self) -> String {
        // Generates a new trace ID for a request batch.
        let mut trace_id = /* uuid */ "00000000-0000-0000-0000-000000000000".to_string().to_string()[..8];
        self.metrics["traces".to_string()][trace_id] = vec![];
        trace_id
    }
    /// Append a message to a specific trace.
    pub fn log_trace(&self, trace_id: String, message: String) -> () {
        // Append a message to a specific trace.
        if !self.metrics["traces".to_string()].contains(&trace_id) {
            return;
        }
        self.metrics["traces".to_string()][&trace_id].push(format!("[{}] {}", time::strftime("%H:%M:%S".to_string()), message));
        logger.debug(format!("[Trace:{}] {}", trace_id, message));
    }
    pub fn add_metric(&self, key: String, value: f64) -> () {
        if !self.metrics.contains(&key) {
            return;
        }
        self.metrics[&key].push(value);
        logger.debug(format!("[Metric] {}: {:.2}", key, value));
    }
    pub fn get_averages(&self) -> () {
        self.metrics.iter().iter().map(|(k, v)| (k, if v { (v.iter().sum::<i64>() / v.len()) } else { 0 })).collect::<HashMap<_, _>>()
    }
}

/// Decorator to measure execution time of a function.
/// Automatically logs results to PerformanceMonitor.
pub fn profile_execution(name: String) -> () {
    // Decorator to measure execution time of a function.
    // Automatically logs results to PerformanceMonitor.
    let decorator = |func| {
        // Decorator.
        let wrapper = || {
            // Wrapper.
            let mut start = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
            let mut result = func(/* *args */, /* ** */ kwargs);
            let mut duration_ms = ((std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start) * 1000);
            if name.to_lowercase().contains(&"rag".to_string()) {
                monitor.add_metric("rag_retrieval".to_string(), duration_ms);
            } else if name.to_lowercase().contains(&"expert".to_string()) {
                monitor.add_metric("expert_latency".to_string(), duration_ms);
            }
            logger.info(format!("[Profiler] {} took {:.1}ms", name, duration_ms));
            result
        };
        wrapper
    };
    decorator
}

/// Async version of the execution profiler.
pub fn profile_async_execution(name: String) -> () {
    // Async version of the execution profiler.
    let decorator = |func| {
        // Decorator.
        let wrapper = || {
            // Wrapper.
            let mut start = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
            let mut result = func(/* *args */, /* ** */ kwargs).await;
            let mut duration_ms = ((std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start) * 1000);
            if name.to_lowercase().contains(&"rag".to_string()) {
                monitor.add_metric("rag_retrieval".to_string(), duration_ms);
            }
            logger.info(format!("[Profiler] {} took {:.1}ms", name, duration_ms));
            result
        };
        wrapper
    };
    decorator
}
